_posts/2024-04-12-vector-search.md

```diff
@@ -0,0 +1,104 @@
+CREATE TABLE embeddings (vec FLOAT[3]);
+
+-- Create an HNSW index on the column
+CREATE INDEX idx ON embeddings USING HNSW (vec);
+```
+
+This index type can't be used to enforce constraints or uniqueness like the built-in `ART` index, and can't be used to speed up joins or index regular columns either. Instead, the `HNSW` index is only applicable to columns of the `ARRAY` type and will only be used to accelerate queries calculating the "distance" between a constant array and the arrays in the indexed column, ordered by the resulting distance and returning the top-n results. That is, queries on the form:
+
+```sql
+SELECT * 
+FROM embeddings
+ORDER BY array_distance(vec, [1,2,3]::FLOAT[3]) 
+LIMIT 5;
+```
+
+...will have their logical plan optimized to instead become a projection over a new `HNSW` index scan operator, removing the limit and sort altogether. We can verify this by checking the `EXPLAIN` output!
+
+```sql
+EXPLAIN SELECT * FROM embeddings ORDER BY array_distance(vec, [1,2,3]::FLOAT[3]) LIMIT 3;
+
+┌───────────────────────────┐
+│         PROJECTION        │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│             #0            │
+└─────────────┬─────────────┘                             
+┌─────────────┴─────────────┐
+│         PROJECTION        │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│            vec            │
+│array_distance(vec, [1.0, 2│
+│         .0, 3.0])         │
+└─────────────┬─────────────┘                             
+┌─────────────┴─────────────┐
+│      HNSW_INDEX_SCAN      │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│   t1 (HNSW INDEX SCAN :   │
+│            idx)           │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│            vec            │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│           EC: 3           │
+└───────────────────────────┘   
+```
+
+You can pass the `HNSW` index creation statement a `metric` parameter to decide what kind of distance metric to use. The supported metrics are `l2sq`, `cosine` and `inner_product`, matching the three built in distance functions: `array_distance`, `array_cosine_distance` and `array_inner_product`.
+ The default is `l2sq` (`array_distance`)..
+
+```sql
+-- Create an HNSW index on the column using euclidean distance
+CREATE INDEX l2sq_idx ON embeddings USING HNSW (vec) WITH (metric = 'l2sq');
+
+-- Create an HNSW index on the column using cosine distance
+CREATE INDEX cos_idx ON embeddings USING HNSW (vec) WITH (metric = 'cosine');
+
+-- Create an HNSW index on the column using inner product
+CREATE INDEX ip_idx ON embeddings USING HNSW (vec) WITH (metric = 'ip');
+```
+
+It is important to know that the `HNSW` index is a _approximate_ index, meaning that the results returned by the index may not be the _exact_ top-n results, but rather a set of results that are _close_ to the top-n results. The `HNSW` in `HNSW` index stands for Hierarchical Navigable Small World, which is a graph-based index structure that is particularly well suited for high-dimensional vector similarity search. 
```

Comment:
I haven't found a good spot in the flow, but this feels a little late to define HNSW. If you can slide it in earlier and keep your nice flow, that would be helpful I think!